### PR TITLE
GH-3903: Improve AOT for gateway proxy beans

### LINKED ISSUE #3903: GatewayProxyBeanRegistrationAotProcessor should not return a CodeFragments
`GatewayProxyBeanRegistrationAotProcessor` currently returns a custom code fragments but, looking at the code, it does not customize any code generation at all.

If you need to register hints on the state of the `BeanFactory`, consider implementing `BeanFactoryInitializationAotProcessor`, see `ReflectiveProcessorBeanFactoryInitializationAotProcessor` for an example.

### PR DESCRIPTION
**Pull request by artembilan**
Fixes https://github.com/spring-projects/spring-integration/issues/3903

* Fix `MessagingGatewayRegistrar` to set a `targetType` on the bean definition instead of `FactoryBean.OBJECT_TYPE_ATTRIBUTE`
* Rework `GatewayProxyBeanRegistrationAotProcessor` to the `GatewayProxyInitializationAotProcessor implements BeanFactoryInitializationAotProcessor` to avoid custom code generation
* Remove tests which rely on the `FactoryBean.OBJECT_TYPE_ATTRIBUTE`

### CODE REVIEW

**Comment by artembilan** [comment]
@snicoll ,

I have figured out that I can get rid off `FactoryBean.OBJECT_TYPE_ATTRIBUTE`  in favor of `setTargetType()`.
Works well in plain JVM.
It fails directly on AOT compilation phase just because the generated code is wrong:
```
 /**
   * Bean definitions for {@link IntegrationApplication.ControlBusGateway}
   */
  public static class ControlBusGateway__BeanDefinitions {
    /**
     * Get the bean instance supplier for 'integrationApplication$ControlBusGateway'.
     */
    private static BeanInstanceSupplier<IntegrationApplication.ControlBusGateway> getControlBusGatewayInstanceSupplier(
        ) {
      return BeanInstanceSupplier.<IntegrationApplication.ControlBusGateway>forConstructor(Class.class)
              .withGenerator((registeredBean, args) -> new GatewayProxyFactoryBean(args.get(0)));
    }

    /**
     * Get the bean definition for 'controlBusGateway'
     */
    public static BeanDefinition getControlBusGatewayBeanDefinition() {
      Class<?> beanType = IntegrationApplication.ControlBusGateway.class;
      RootBeanDefinition beanDefinition = new RootBeanDefinition(beanType);
      beanDefinition.getConstructorArgumentValues().addIndexedArgumentValue(0, IntegrationApplication.ControlBusGateway.class);
      beanDefinition.getPropertyValues().addPropertyValue("defaultRequestChannelName", "controlBus.input");
      beanDefinition.getPropertyValues().addPropertyValue("proxyDefaultMethods", "false");
      beanDefinition.getPropertyValues().addPropertyValue("defaultRequestTimeoutExpressionString", "-9223372036854775808");
      beanDefinition.getPropertyValues().addPropertyValue("defaultReplyTimeoutExpressionString", "-9223372036854775808");
      beanDefinition.getPropertyValues().addPropertyValue("methodMetadataMap", null);
      beanDefinition.setInstanceSupplier(getControlBusGatewayInstanceSupplier());
      return beanDefinition;
    }
  }
```

The `beanType` must be a `GatewayProxyFactoryBean`.

So, I leave this as a `Draft` for future consideration since there is indeed no good support for `FactoryBean` in AOT mode.


The previous version, based on the `FactoryBean.OBJECT_TYPE_ATTRIBUTE` looks like this:
```
  /**
   * Get the bean instance supplier for 'integrationApplication$ControlBusGateway'.
   */
  private static BeanInstanceSupplier<GatewayProxyFactoryBean> getControlBusGatewayInstanceSupplier(
      ) {
    return BeanInstanceSupplier.<GatewayProxyFactoryBean>forConstructor(Class.class)
            .withGenerator((registeredBean, args) -> new GatewayProxyFactoryBean(args.get(0)));
  }

  /**
   * Get the bean definition for 'controlBusGateway'
   */
  private static BeanDefinition getControlBusGatewayBeanDefinition() {
    Class<?> beanType = GatewayProxyFactoryBean.class;
    RootBeanDefinition beanDefinition = new RootBeanDefinition(beanType);
    beanDefinition.getConstructorArgumentValues().addIndexedArgumentValue(0, IntegrationApplication.ControlBusGateway.class);
    beanDefinition.getPropertyValues().addPropertyValue("defaultRequestChannelName", "controlBus.input");
    beanDefinition.getPropertyValues().addPropertyValue("proxyDefaultMethods", "false");
    beanDefinition.getPropertyValues().addPropertyValue("defaultRequestTimeoutExpressionString", "-9223372036854775808");
    beanDefinition.getPropertyValues().addPropertyValue("defaultReplyTimeoutExpressionString", "-9223372036854775808");
    beanDefinition.getPropertyValues().addPropertyValue("methodMetadataMap", null);
    beanDefinition.setAttribute("factoryBeanObjectType", IntegrationApplication.ControlBusGateway.class);
    beanDefinition.setInstanceSupplier(getControlBusGatewayInstanceSupplier());
    return beanDefinition;
  }
```

Thanks for understanding

File: spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java [hunk 1]
```diff
@@ -183,8 +180,8 @@ else if (StringUtils.hasText(asyncExecutor)) {
 
 		gatewayProxyBuilder.addConstructorArgValue(serviceInterface);
 
-		AbstractBeanDefinition beanDefinition = gatewayProxyBuilder.getBeanDefinition();
-		beanDefinition.setAttribute(FactoryBean.OBJECT_TYPE_ATTRIBUTE, serviceInterface);
+		RootBeanDefinition beanDefinition = (RootBeanDefinition) gatewayProxyBuilder.getBeanDefinition();
+		beanDefinition.setTargetType(serviceInterface);
```
Comment (snicoll) [inline spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java:184]: You also need to call `setBeanClass` with the factory bean. It is tricky and one of those areas where the core container is not fully consistent. 

Give that a try and let me know how it goes.

Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java:184]: Didn't help, Stephane.

My code is like this currently:
```
		RootBeanDefinition beanDefinition = (RootBeanDefinition) gatewayProxyBuilder.getBeanDefinition();
		beanDefinition.setBeanClass(GatewayProxyFactoryBean.class);
		beanDefinition.setTargetType(serviceInterface);
```
But an AOT Smoke still fails with the same error: a generated `getControlBusGatewayBeanDefinition()` content is without any changes so far.
I pulled latest `main` and used this command to verify:
```
gradlew clean :integration:build --include-build=..\spring-integration
```
Where `spring-integration` is exactly on this branch with the change we have just discussed.

**Comment by artembilan** [comment]
It looks like with generic argument we now expose on the `ProxyFactoryBean`, it works with AOT even without the mentioned `FactoryBean.OBJECT_TYPE_ATTRIBUTE` and without custom code.
In the end it looks like AOT generates a correct working code:
```
    private static BeanInstanceSupplier<GatewayProxyFactoryBean> getControlBusGatewayInstanceSupplier(
        ) {
      return BeanInstanceSupplier.<GatewayProxyFactoryBean>forConstructor(Class.class)
              .withGenerator((registeredBean, args) -> new GatewayProxyFactoryBean(args.get(0)));
    }

    public static BeanDefinition getControlBusGatewayBeanDefinition() {
      ResolvableType beanType = ResolvableType.forClassWithGenerics(GatewayProxyFactoryBean.class, IntegrationApplication.ControlBusGateway.class);
      RootBeanDefinition beanDefinition = new RootBeanDefinition(beanType);
      beanDefinition.getConstructorArgumentValues().addIndexedArgumentValue(0, IntegrationApplication.ControlBusGateway.class);
      beanDefinition.getPropertyValues().addPropertyValue("defaultRequestChannelName", "controlBus.input");
      beanDefinition.getPropertyValues().addPropertyValue("proxyDefaultMethods", "false");
      beanDefinition.getPropertyValues().addPropertyValue("defaultRequestTimeoutExpressionString", "-9223372036854775808");
      beanDefinition.getPropertyValues().addPropertyValue("defaultReplyTimeoutExpressionString", "-9223372036854775808");
      beanDefinition.getPropertyValues().addPropertyValue("methodMetadataMap", null);
      beanDefinition.setInstanceSupplier(getControlBusGatewayInstanceSupplier());
      return beanDefinition;
    }
```

**Pull request merged**